wasmtime-api/src/externals.rs

```diff
@@ -411,7 +434,14 @@ impl Memory {
         }
     }
 
-    pub fn data(&self) -> *mut u8 {
+    pub fn data(&self) -> &mut [u8] {
+        unsafe {
+            let definition = &*self.wasmtime_memory_definition();
+            slice::from_raw_parts_mut(definition.base, definition.current_length)
+        }
+    }
```

Comment:
Handing out a Rust reference to the memory is potentially unsafe; if some other thread does a `memory.grow` while this reference is live, it could dangle. I filed https://github.com/WebAssembly/wasm-c-api/issues/105 to track the same issue in wasm-c-api.

But regardless of whether wasm-c-api itself changes, that issue makes this a potentially unsafe function. Perhaps we could add `unsafe` here? And/or add load/store functions for accessing data without handing out a slice?